validating admission webhook for storageclass

divyenpatel opened the pull request:
<!-- Thanks for sending a pull request! -->

**What this PR does / why we need it**:
We need this PR to meet following requirements

1. In the translation library for vSphere CSI Migration, we have added migration specific parameters to help send VCP parameters to CSI Driver. We are documenting that, new storage class should not be created using these migration specific parameters. We want to prevent a user from creating a storage class with these migration specific parameters. Refer: https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/312#discussion_r468377097

2. vSphere Cloud Provider was not supporting volume expansion, but vSphere CSI Driver does. We want to prevent the user to expand the capability of in-tree volume. For now, we are returning an un-implemented response when the migrated volume is observed in the Controller expand volume call. Upstream reviewers suggested writing an admission controller to prevent a user from adding "allowVolumeExpansion: true" in-tree VCP storage class. 
Refer: https://github.com/kubernetes/kubernetes/issues/93727#issuecomment-669710189


This PR is adding 
- script for generating k8s singed certificates and certificate secret volume
- script for creating validation webhook service - `vsphere-webhook-svc` , `ValidatingWebhookConfiguration`, webhook deployment, and service account, roles and role binding for webhook deployment pod.
- adding operation modes to syncer container - `WEBHOOK_SERVER` and `METADATA_SYNC`. modes can be configured via a command-line flag - `operation-mode`
-  admission webhook handler for storage class validations.
-  added support for webhook server certificate rotation.
- unit tests for admission review validations.

 

**Which issue this PR fixes**
fixes https://github.com/kubernetes-sigs/vsphere-csi-driver/issues/317

**Special notes for your reviewer**:
Executed full sync and metadata syncer tests to ensure we are not breaking existing functionality. 

```
export GINKGO_FOCUS="full-sync-test|label-updates\sfor\sfile\svolumes"
```

Test Logs: 
[e2e-full-sync-pr-337.log](https://github.com/kubernetes-sigs/vsphere-csi-driver/files/5258514/e2e-full-sync-pr-337.log)


```
Ran 8 of 106 Specs in 2948.793 seconds
SUCCESS! -- 8 Passed | 0 Failed | 0 Pending | 98 Skipped
PASS

Ginkgo ran 1 suite in 49m24.468512616s
Test Suite Passed
```
**Unit Tests**

```
=== RUN   TestValidateStorageClassForAllowVolumeExpansion
{"level":"info","time":"2020-09-12T00:49:30.288605-07:00","caller":"admissionhandler/validatestorageclass.go:53","msg":"Validating StorageClass: \"sc\""}
{"level":"info","time":"2020-09-12T00:49:30.289978-07:00","caller":"admissionhandler/validatestorageclass.go:78","msg":"Validation of StorageClass: \"sc\" Failed"}
--- PASS: TestValidateStorageClassForAllowVolumeExpansion (0.01s)
    validatestorageclass_test.go:49: TestValidateStorageClassForAllowVolumeExpansion Passed
=== RUN   TestValidateStorageClassForMigrationParameter
{"level":"info","time":"2020-09-12T00:49:30.292081-07:00","caller":"admissionhandler/validatestorageclass.go:53","msg":"Validating StorageClass: \"sc\""}
{"level":"info","time":"2020-09-12T00:49:30.292117-07:00","caller":"admissionhandler/validatestorageclass.go:78","msg":"Validation of StorageClass: \"sc\" Failed"}
--- PASS: TestValidateStorageClassForMigrationParameter (0.00s)
    validatestorageclass_test.go:64: TestValidateStorageClassForMigrationParameter Passed
=== RUN   TestValidateStorageClassForValidStorageClass
{"level":"info","time":"2020-09-12T00:49:30.293419-07:00","caller":"admissionhandler/validatestorageclass.go:53","msg":"Validating StorageClass: \"sc\""}
{"level":"info","time":"2020-09-12T00:49:30.293443-07:00","caller":"admissionhandler/validatestorageclass.go:76","msg":"Validation of StorageClass: \"sc\" Passed"}
--- PASS: TestValidateStorageClassForValidStorageClass (0.00s)
    validatestorageclass_test.go:79: TestValidateStorageClassForValidStorageClass Passed
PASS
ok  	sigs.k8s.io/vsphere-csi-driver/pkg/syncer/admissionhandler	1.161s
```

**Deployment**

vSphere CSI webhook Pod runs in seperate deployment.

```
% kubectl get deployment vsphere-csi-webhook --namespace=kube-system
NAME                  READY   UP-TO-DATE   AVAILABLE   AGE
vsphere-csi-webhook   1/1     1            1           5m21s
```

```
 ./generate-signed-webhook-certs.sh 
creating certs in tmpdir /var/folders/vy/_6dvxx7j5db9sq9n38qjymwr002gzv/T/tmp.mclIK6Jn 
Generating RSA private key, 2048 bit long modulus
..............................+++
..................................................................................................+++
e is 65537 (0x10001)
certificatesigningrequest.certificates.k8s.io "vsphere-webhook-svc.kube-system" deleted
Warning: certificates.k8s.io/v1beta1 CertificateSigningRequest is deprecated in v1.19+, unavailable in v1.22+; use certificates.k8s.io/v1 CertificateSigningRequest
certificatesigningrequest.certificates.k8s.io/vsphere-webhook-svc.kube-system created
NAME                              AGE   SIGNERNAME                     REQUESTOR          CONDITION
vsphere-webhook-svc.kube-system   0s    kubernetes.io/legacy-unknown   kubernetes-admin   Pending
certificatesigningrequest.certificates.k8s.io/vsphere-webhook-svc.kube-system approved
secret/vsphere-webhook-certs configured
```


```
./create-validation-webhook.sh
service "vsphere-webhook-svc" deleted
validatingwebhookconfiguration.admissionregistration.k8s.io "validation.csi.vsphere.vmware.com" deleted
serviceaccount "vsphere-csi-webhook" deleted
clusterrole.rbac.authorization.k8s.io "vsphere-csi-webhook-role" deleted
clusterrolebinding.rbac.authorization.k8s.io "vsphere-csi-webhook-role-binding" deleted
deployment.apps "vsphere-csi-webhook" deleted
service/vsphere-webhook-svc created
validatingwebhookconfiguration.admissionregistration.k8s.io/validation.csi.vsphere.vmware.com created
serviceaccount/vsphere-csi-webhook created
clusterrole.rbac.authorization.k8s.io/vsphere-csi-webhook-role created
clusterrolebinding.rbac.authorization.k8s.io/vsphere-csi-webhook-role-binding created
deployment.apps/vsphere-csi-webhook created
```



After Executing the above two scripts webhook configuration, webhook secret and webhook service and webhook deployment will be created.

webhook secret looks like as below

<img width="791" alt="Screen Shot 2020-09-17 at 11 41 55 PM" src="https://user-images.githubusercontent.com/22985595/93565029-881bca80-f93f-11ea-93f3-790b0d2ccf52.png">


`vsphere-webhook-svc` service binds to `vsphere-csi-webhook` pod.

```
% kubectl describe service vsphere-webhook-svc --namespace=kube-system
Name:              vsphere-webhook-svc
Namespace:         kube-system
Labels:            app=vsphere-csi-webhook
Annotations:       <none>
Selector:          app=vsphere-csi-webhook
Type:              ClusterIP
IP:                10.108.246.80
Port:              <unset>  443/TCP
TargetPort:        8443/TCP
Endpoints:         10.244.0.71:8443
Session Affinity:  None
Events:            <none>
```



**Testing Done**

Performed Certificate Rotation After the webhook server was up and running.

```
{"level":"info","time":"2020-09-18T06:38:57.130878949Z","caller":"admissionhandler/admissionhandler.go:80","msg":"Restarting webhook server with new config","TraceId":"60ac49c3-f6b1-4593-9d0a-d92371c7c06f"}
{"level":"info","time":"2020-09-18T06:38:57.134094286Z","caller":"admissionhandler/admissionhandler.go:178","msg":"shutting down webhook server gracefully...","TraceId":"60ac49c3-f6b1-4593-9d0a-d92371c7c06f"}
{"level":"info","time":"2020-09-18T06:38:57.135490729Z","caller":"admissionhandler/admissionhandler.go:184","msg":"restarting webhook server...","TraceId":"60ac49c3-f6b1-4593-9d0a-d92371c7c06f"}
{"level":"info","time":"2020-09-18T06:38:57.136872035Z","caller":"admissionhandler/admissionhandler.go:156","msg":"webhook server stopped","TraceId":"70f1f735-eff2-45b8-bd44-69c71348d4c1"}
{"level":"info","time":"2020-09-18T06:38:57.137622506Z","caller":"admissionhandler/admissionhandler.go:162","msg":"webhook server started","TraceId":"60ac49c3-f6b1-4593-9d0a-d92371c7c06f"}
```

User trying to create a new CSI storageclass with migration specific parameters.

```
# kubectl create -f sc.yaml 
Error from server (Invalid StorageClass Parameters. Migration specific parameters should not be used in the storage class): error when creating "sc.yaml": admission webhook "scvalidation.csi.vsphere.vmware.com" denied the request: Invalid StorageClass Parameters. Migration specific parameters should not be used in the storage class
```

User trying to create a new CSI storageclass with migration specific parameters.

```
# kubectl create -f sc.yaml
Error from server (AllowVolumeExpansion can not be set to true on the in-tree vSphere storage class): error when creating "sc-volume-expansion.yaml": admission webhook "scvalidation.csi.vsphere.vmware.com" denied the request: AllowVolumeExpansion can not be set to true on the in-tree vSphere storage class
```

User trying to edit a VCP storageclass with allowVolumeExpansion: true

```
# kubectl edit sc vcpsc
Edit cancelled, no changes made.
```

User trying to patch a VCP storageclass with allowVolumeExpansion: true

```
# kubectl apply  -f sc-volume-expansion.yaml 
Error from server (AllowVolumeExpansion can not be set to true on the in-tree vSphere storage class): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"allowVolumeExpansion\":true,\"apiVersion\":\"storage.k8s.io/v1\",\"kind\":\"StorageClass\",\"metadata\":{\"annotations\":{},\"name\":\"vcpsc\"},\"parameters\":{\"hostFailuresToTolerate\":\"2\"},\"provisioner\":\"kubernetes.io/vsphere-volume\"}\n"}}}
to:
Resource: "storage.k8s.io/v1, Resource=storageclasses", GroupVersionKind: "storage.k8s.io/v1, Kind=StorageClass"
Name: "vcpsc", Namespace: ""
for: "sc-volume-expansion.yaml": admission webhook "scvalidation.csi.vsphere.vmware.com" denied the request: AllowVolumeExpansion can not be set to true on the in-tree vSphere storage class
```


**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
validating admission webhook for storageclass
```

cc: @chethanv28 @SandeepPissay @xing-yang

svcbot-qecnsdp commented:
Can one of the admins verify this patch?

andrewsykim commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+        operations:  ["CREATE", "UPDATE"]
+        resources:   ["storageclasses"]
+    sideEffects: None
+    admissionReviewVersions: ["v1", "v1beta1"]
```
Here we're receiving v1beta1 AdminssionReview, but I think the webhook server is only decoding v1 types.

divyenpatel replied:
we will be configuring webhook on k8s cluster 1.19 or above.

k8s 1.19 already supports `admissionregistration.k8s.io/v1`. I guess we can remove `v1beta1` from the list.
```
% kubectl api-versions | grep admissionregistration
admissionregistration.k8s.io/v1
admissionregistration.k8s.io/v1beta1
```

RaunakShah commented on cmd/syncer/main.go:
```diff
@@ -65,6 +68,12 @@ func main() {
 		os.Exit(1)
 	}
+	// Initialize Kubernetes feature state config instance
+	k8sFeatureStateConfig, err := commonco.GetContainerOrchestratorInterface(ctx, common.Kubernetes, configInfo.Cfg.FeatureStatesConfig)
```
shouldn't this be within the `run` function? We only want one instance of the admission webhook to be the leader right?

divyenpatel replied:
No, we want all webhook instances to be up and running regardless of leader election. 
If we create multiple replicas of our controller pod, the service bind to all those pods and if a validation request comes to the pod where the webhook is not running, the operation will be disallowed because the webhook is not accessible.

BaluDontu replied:
>If we create multiple replicas of our controller pod
But this is not a supported model right for our CSI controller?

>  if a validation request comes to the pod where the webhook is not running, the operation will be disallowed because the webhook is not accessible.
If guarded by leader election, the validation request will only go to the webhook where the leader is running. isn't that true?

divyenpatel replied:
> If guarded by leader election, the validation request will only go to the webhook where the leader is running. isn't that true?

No this is not true. The validation request does not honor or look who is a leader. 
A review request can be sent to any endpoint address bound to the service.

RaunakShah commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/generate-signed-webhook-certs.sh:
```diff
@@ -0,0 +1,100 @@
+
+service=scvalidation-webhook-svc
+secret=scvalidation-webhook-certs
+namespace=kube-system
```
what if csi driver is not installed in `kube-system` namespace?

divyenpatel replied:
all of our YAML files are currently uses `kube-system` we need to create service in the namespace where controller pod is running.

if we are deploying driver in the different namespace we need to change this. I can supply this as an argument and default to `kube-system` if not specified anything to script.

RaunakShah commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,207 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common"
+
```
nit: can remove the empty line

divyenpatel replied:
done.

RaunakShah commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,207 @@
+	webhookServerPort     = 8443
+	server                *http.Server
+	unSupportedParameters = parameterSet{
+		common.CSIMigrationParams:                   emptyStruct{},
```
for my understanding: `emptyStruct` is of type `struct{}` so `emptyStruct{}` is of type `struct{}{}`?

divyenpatel replied:
yes. and I can replace `emptyStruct{}` with `struct{}{}`

RaunakShah commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,207 @@
+// StartWebhookServer starts the webhook server
+func StartWebhookServer(ctx context.Context) error {
+	log := logger.GetLogger(ctx)
+	certs, err := tls.LoadX509KeyPair("/etc/certs/cert.pem", "/etc/certs/key.pem")
```
Maybe in a follow up change - Is this something we want to read from the YAML? The path doesn't necessarily have to be `/etc/certs` right?

divyenpatel replied:
I am planning to supply this in the `vsphere.conf`file, so we do not need additional secret file to be mounted on the pod.

BaluDontu replied:
can you add a TODO there?

divyenpatel replied:
@BaluDontu I have already addressed this comment - please take a look at the latest commit - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/337/commits/f5788a0bb8163cf2a51c6fbd4adbfdf263c81b0c

RaunakShah commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,207 @@
+		return err
+	}
+	server = &http.Server{
+		Addr:      fmt.Sprintf(":%v", webhookServerPort),
```
Same with `webhookServerPort`, maybe something that can be configurable?

divyenpatel replied:
let me supply this as well in the vsphere.conf file.

RaunakShah commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,207 @@
+		}
+	}
+	if len(body) == 0 {
+		log.Error("empty body")
```
More descriptive error message?

divyenpatel replied:
changed it to 

```
		log.Error("received empty request body")
		http.Error(w, "received empty request body", http.StatusBadRequest)
```

RaunakShah commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,207 @@
+			}
+		}
+	}
+	// return AdmissionResponse result
```
Can we have an info level log here stating that the storage class was unsuccessfully/successfully validated?

divyenpatel replied:
added info level log

```
log.Infof("Validation of StorageClass: %q Passed", sc.Name)
```

```
log.Infof("Validation of StorageClass: %q Failed", sc.Name)
```

RaunakShah commented:
Can you run the e2e pipeline?

sashrith commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+    rules:
+      - apiGroups:   ["storage.k8s.io"]
+        apiVersions: ["v1", "v1beta1"]
+        operations:  ["CREATE", "UPDATE"]
```
what about `patch`

divyenpatel replied:
patch is not supported value. so update will be considered as patch.

> Unsupported value: "PATCH": supported values: "*", "CONNECT", "CREATE", "DELETE", "UPDATE"

gohilankit commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: sc-validation-webhook-svc
```
The 'sc' prefix can be removed from the name, as the same service can be used for all admission validations, not only storage classes.

BaluDontu replied:
+1

divyenpatel replied:
fixed.

gohilankit commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/create-validation-webhook.sh:
```diff
@@ -0,0 +1,20 @@
+CA_BUNDLE=$(kubectl get configmap -n kube-system extension-apiserver-authentication -o=jsonpath='{.data.client-ca-file}' | base64 | tr -d '\n')
+
+# clean-up previously created service and validatingwebhookconfiguration. Ignore errors if not present.
+kubectl delete validatingwebhookconfiguration.admissionregistration.k8s.io scvalidation.csi.vsphere.vmware.com 2>/dev/null || true
```
The resource name for validatingwebhookconfiguration & the service should match the names in validatingwebhook.yaml

divyenpatel replied:
done.

BaluDontu commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+  - name: scvalidation.csi.vsphere.vmware.com
+    clientConfig:
+      service:
+        name: sc-validation-webhook-svc
```
remove sc from here. Let's have a single validation webhook for everything.

divyenpatel replied:
service name is now set to `validation-webhook-svc`

BaluDontu commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+metadata:
+  name: validation.csi.vsphere.vmware.com
+webhooks:
+  - name: scvalidation.csi.vsphere.vmware.com
```
Lets name it validation.csi.vsphere.vmware.com. Lets not make it specific to sc.

divyenpatel replied:
webhook name is set to `validation.csi.vsphere.vmware.com`

BaluDontu commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+      service:
+        name: sc-validation-webhook-svc
+        namespace: kube-system
+        path: "/validate-storageclass"
```
can we keep a generic name for path and we can have multiple rules if needed here.

divyenpatel replied:
used validate as the path to help validate all diff types of k8s resources.

```
    clientConfig:
      service:
        name: validation-webhook-svc
        namespace: kube-system
        path: "/validate"
```

BaluDontu commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,142 @@
+			},
+		}
+	} else {
+		if r.URL.Path == "/validate-storageclass" {
```
lets keep this generic path. Once you decode, determine if it is a storage class object and call validateStorageClass.

divyenpatel replied:
done.

```
              if r.URL.Path == "/validate" {
			log.Debugf("request URL path is /validate")
			log.Debugf("admissionReview: %+v", ar)
			switch ar.Request.Kind.Kind {
			case "StorageClass":
				admissionResponse = validateStorageClass(ctx, &ar)
			default:
				admissionResponse = &admissionv1.AdmissionResponse{
					Allowed: true,
				}
			}
			log.Debugf("admissionResponse: %+v", admissionResponse)
		}
```

BaluDontu commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/create-validation-webhook.sh:
```diff
@@ -0,0 +1,20 @@
+if [ "$1" = "-h" ] || [ "$1" = "--help" ]; then
+  cat <<EOF
+Usage: Patch validatingwebhook.yaml with CA_BUNDLE retrieved from Kubernetes API server
+and create ValidatingWebhookConfiguration and scvalidation-webhook-svc service using patched yaml file
```
Lets use service name as validation-webhook-svc

divyenpatel replied:
fixed.

divyenpatel commented:
@gohilankit @BaluDontu I have addressed your review comments here https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/337/commits/a0ecfd952491d618f6c9a8c1efb1bb93cac18b04

PTAL

gohilankit commented:
> @gohilankit @BaluDontu I have addressed your review comments here [a0ecfd9](https://github.com/kubernetes-sigs/vsphere-csi-driver/commit/a0ecfd952491d618f6c9a8c1efb1bb93cac18b04)
> 
> PTAL

LGTM. One question - How are we planning to update webhook certs in the secret for supervisor and guest cluster flavors? I believe the webhook server needs to be run for all the flavors.

BaluDontu commented:
@gohilankit I don't think we would certs for webhook in SV. Our webhook would be running as part of CSI pod would be running on master and nobody would have access to the master VMs in SV.

For TKG, I guess we might need one as SV dev ops would have full access to the TKG.

gohilankit commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,148 @@
+}
+
+var (
+	webhookServerPort = 8443
```
Can we change the default port to something else (say 9443), as 8443 is already being used for Authproxy in 7.0u1 ?

BaluDontu replied:
Can we make these ports configurable?

gohilankit replied:
It is configurable. If the admission-webhook-port value is not supplied in config, then it picks up this default value.

divyenpatel replied:
@gohilankit we have not used port `8443` in our pod anywhere. If `8443` is used in the other pod I think we are good.
As we bind our service to our controller pod only. And also this port can be configured using vSphere config secret.

BaluDontu commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,148 @@
+	}
+	server = &http.Server{
+		Addr:      fmt.Sprintf(":%v", port),
+		TLSConfig: &tls.Config{Certificates: []tls.Certificate{certs}},
```
Can we add this TLSConfig only if certs are non-empty. For SV, we don't need to run webhook server securely as CNS-CSI controller is running on master which is not accessible to any user.

divyenpatel replied:
This is not true Balu. SV also starts a webhook server using a certificate in the scheduler extension.
In SV  ValidatingWebhookConfiguration uses local address `127.0.0.1`, and thus does not require to supply CABundle in the `ValidatingWebhookConfiguration`.

```
clientConfig:
    url: 'https://127.0.0.1:12345/scheduler/admission'
```

SandeepPissay commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,257 @@
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		log.Errorf("Content-Type=%s, expect application/json", contentType)
```
Nit: Fix error message to not start with init cap.

divyenpatel replied:
done.

SandeepPissay commented on pkg/syncer/admissionhandler/validatestorageclass.go:
```diff
@@ -0,0 +1,108 @@
+	allowed := true
+
+	switch req.Kind.Kind {
+	case "StorageClass":
```
We need code to fail validation if the kind is not StorageClass.

divyenpatel replied:
added default case.

```
switch req.Kind.Kind {
	case "StorageClass":
        .
        .
        default:
		allowed = false
		log.Errorf("Can't validate resource kind: %q using validateStorageClass function", req.Kind.Kind)
	}
}
```

SandeepPissay commented on pkg/syncer/admissionhandler/validatestorageclass.go:
```diff
@@ -0,0 +1,108 @@
+			}
+		}
+		// Migration parameters check for csi.vsphere.vmware.com provisioner
+		if allowed && sc.Provisioner == "csi.vsphere.vmware.com" {
```
Can we move this to else if block?

divyenpatel replied:
done.

SandeepPissay commented on pkg/syncer/admissionhandler/validatestorageclass.go:
```diff
@@ -0,0 +1,108 @@
+		if allowed {
+			log.Infof("Validation of StorageClass: %q Passed", sc.Name)
+		} else {
+			log.Infof("Validation of StorageClass: %q Failed", sc.Name)
```
Infof -> Errorf

divyenpatel replied:
done.

SandeepPissay reviewed (commented):
Webhook availability is super critical, else it will bring down the cluster. Can we run the vsphere-webhook in a separate deployment so that we have comppletely decoupled it from CSI pod?

divyenpatel commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/generate-signed-webhook-certs.sh:
```diff
@@ -0,0 +1,161 @@
+while true; do
+    if kubectl get csr "${csrName}"; then
+        break
+    fi
```
done.

```
while true; do
    if kubectl get csr "${csrName}"; then
        break
    fi
    echo "waiting for CertificateSigningRequest: ${csrName} to be available"
    sleep 1
done
```

divyenpatel commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/generate-signed-webhook-certs.sh:
```diff
@@ -0,0 +1,161 @@
+    if [[ ${serverCert} != '' ]]; then
+        break
+    fi
+    sleep 1
```
done.

```
for _ in $(seq 10); do
    serverCert=$(kubectl get csr ${csrName} -o jsonpath='{.status.certificate}')
    if [[ ${serverCert} != '' ]]; then
        break
    fi
    echo "waiting for certificate request to complete"
    sleep 1
done
```

divyenpatel commented on pkg/syncer/admissionhandler/admissionhandler.go:
```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		log.Fatalf("failed to get webhook config. err: %v", err)
+	}
+	log.Debugf("webhook config: %v", cfg)
```
done.

divyenpatel commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/validatingwebhook.yaml:
```diff
@@ -0,0 +1,35 @@
+  name: vsphere-webhook-svc
+  namespace: kube-system
+  labels:
+    app: vsphere-csi-controller
```
done.

```
kind: Deployment
apiVersion: apps/v1
metadata:
  name: vsphere-csi-webhook
  namespace: kube-system
spec:
  replicas: 1
  selector:
    matchLabels:
      app: vsphere-csi-webhook
  template:
    metadata:
      labels:
        app: vsphere-csi-webhook
        role: vsphere-csi-webhook
```

divyenpatel commented on manifests/latest/vsphere-7.0u1/vanilla/deploy/vsphere-csi-controller-deployment.yaml:
```diff
@@ -117,12 +117,31 @@ spec:
             - mountPath: /csi
               name: socket-dir
+        # needed only when csi-migration is enabled. Requires k8s 1.19+
+        #- name: vsphere-webhook
```
added separate deployment for webhook.

SandeepPissay commented:
/approve

BaluDontu commented:
/approve

BaluDontu commented:
/lgtm
